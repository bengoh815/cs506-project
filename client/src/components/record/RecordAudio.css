#record-group {
  padding: 1rem;
}

#record-button {
  margin: 0.5rem;
}

#upload-button {
  margin: 0.5rem;
}
