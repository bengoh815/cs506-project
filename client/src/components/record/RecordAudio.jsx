/******************************************************************************
 * Filename: RecordButton.jsx
 * Purpose:  A component that allows the user to record audio.
 * Author:   Victor Nguyen & Don Ma
 *
 * Description:
 * This file contains the RecordButton component that allows the user to record
 * audio. The component displays a button that the user can click to start and
 * stop recording audio. The recorded audio is displayed to the user once the
 * recording has stopped and can be played back and downloaded.
 *
 * Usage:
 * To use this component, import it into the desired file and render it.
 *
 ******************************************************************************/

import React, { useState, useRef } from "react";
import { Button } from "react-bootstrap";
import Playback from "./PlaybackAudio";
import "./RecordAudio.css";

const RecordButton = () => {
  // State to store whether the user is currently recording or not
  const [isRecording, setIsRecording] = useState(false);

  // State to store the MediaRecorder object
  const [mediaRecorder, setMediaRecorder] = useState(null);

  // State to store the recorded audio
  const [recordedAudio, setRecordedAudio] = useState(null);

  // State to store the stream
  const streamRef = useRef(null);

  // State to store the label of the record button
  const [recordButtonLabel, setRecordButtonLabel] = useState("Start Recording");

  /**
   * Requests access to the user's microphone. If access is granted, the MediaRecorder object is created and the recording starts.
   */
  const startRecording = () => {
    if (!navigator.mediaDevices) {
      console.error("User media not available");
      return;
    }

    // Reset the recorded audio
    setRecordButtonLabel("Stop Recording");
    setRecordedAudio(null);

    navigator.mediaDevices
      .getUserMedia({ audio: true })
      .then((stream) => {
        streamRef.current = stream; // Assign the stream to the ref
        const recorder = new MediaRecorder(stream);
        setMediaRecorder(recorder);

        recorder.start();
        setIsRecording(true);

        recorder.ondataavailable = (e) => {
          // e.data contains the audio data

          // Create a new Blob object and create an object URL for it
          const blob = new Blob([e.data], { type: "audio/webm" });
          const url = URL.createObjectURL(blob);

          setRecordedAudio(url);
        };

        recorder.onerror = (e) => console.error(e.error);
      })
      .catch((error) => {
        console.error(error);

        // Alert the user if they denied microphone access and prompt them to give permission in order to record audio
        if (error.name === "NotAllowedError") {
          alert(
            "Microphone permission denied. You need to give permission to record audio",
          );
        }
      });
  };

  /**
   * Stops the recording.
   */
  const stopRecording = () => {
    if (mediaRecorder) {
      setIsRecording(false);
      setRecordButtonLabel("Start Another Recording");
      mediaRecorder.stop();
      streamRef.current.getTracks().forEach((track) => track.stop());
    }
  };

  /**
   * Downloads the recording webm file to the user's device.
   */
  const handleDownload = () => {
    const link = document.createElement("a");
    link.href = recordedAudio;
    link.download = "melody-mapper-recording.webm";
    link.click();
  };

  return (
    <div id="record-group" data-testid="record-group">
      {/* Heading */}
      <h3 id="record-heading" data-testid="record-heading">
        Record
      </h3>

      {/* Record Button */}
      <div id="button-group" style={{ display: "flex", style: {} }}>
        <Button
          id="record-button"
          className="button"
          data-testid="record-button"
          onClick={isRecording ? stopRecording : startRecording}
          variant={isRecording ? "danger" : "success"}
        >
          {recordButtonLabel}
        </Button>

        {/* Currently Recording Indicator */}
        {isRecording && <div className="recording-indicator"></div>}

        {/* Download Recorded Audio Button */}
        {recordedAudio && (
          <Button
            id="download-button"
            className="button"
            data-testid="download-button"
            onClick={handleDownload}
            variant="dark"
          >
            Download Recording
          </Button>
        )}
      </div>

      {/* Recorded Audio */}
      {recordedAudio && <Playback audioSrc={recordedAudio} />}
    </div>
  );
};

export default RecordButton;
