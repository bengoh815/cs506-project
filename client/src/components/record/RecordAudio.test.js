/******************************************************************************
 * Filename: RecordButton.test.js
 * Purpose:  Tests the RecordButton component.
 * Author:   Victor Nguyen
 *
 * Description:
 * This file contains tests for the RecordButton component. The tests ensure that
 * the component and its elements render correctly. A snapshot tests is included
 * to catch unintended changes to the component's structure or behavior.
 *
 * Usage:
 * Run the tests using the command `npm test`.
 *
 * Note:
 * More tests to be added/implemented for:
 * - Testing the recording functionality
 * - Testing the playback functionality
 * - Testing uploading the recorded audio
 *
 ******************************************************************************/

import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import RecordAudio from "./RecordAudio";

test("whole record button component renders", () => {
  // Arrange
  render(<RecordAudio />);
  const recordButtonElement = screen.getByTestId("record-group");

  // Assert
  expect(recordButtonElement).toBeInTheDocument();
});

test("record heading component renders", () => {
  // Arrange
  render(<RecordAudio />);
  const recordHeadingElement = screen.getByTestId("record-heading");

  // Assert
  expect(recordHeadingElement).toBeInTheDocument();
});

test("correct record heading content", () => {
  // Arrange
  render(<RecordAudio />);
  const recordHeadingElement = screen.getByTestId("record-heading");

  // Assert
  expect(recordHeadingElement).toHaveTextContent("Record");
});

test("record button component renders", () => {
  // Arrange
  render(<RecordAudio />);
  const recordButton = screen.getByTestId("record-button");

  // Assert
  expect(recordButton).toBeInTheDocument();
});

test("correct record button content - start recording", () => {
  // Arrange
  render(<RecordAudio />);
  const recordButton = screen.getByTestId("record-button");

  // Assert
  expect(recordButton).toHaveTextContent("Start Recording");
});

test.skip("correct record button content - stop recording", () => {
  // TODO: Implement test to check if the button content changes to "Stop Recording" when clicked
});

// Snapshot test
test("matches snapshot", () => {
  // Arrange
  const tree = renderer.create(<RecordAudio />).toJSON();

  // Assert
  expect(tree).toMatchSnapshot();
});
