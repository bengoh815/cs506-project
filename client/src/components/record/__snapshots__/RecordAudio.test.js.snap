// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  data-testid="record-group"
  id="record-group"
>
  <h3
    data-testid="record-heading"
    id="record-heading"
  >
    Record
  </h3>
  <button
    className="btn btn-success"
    data-testid="record-button"
    disabled={false}
    id="record-button"
    onClick={[Function]}
    type="button"
  >
    Start Recording
  </button>
</div>
`;
